packages/api/src/globalContext.ts

```diff
@@ -1,13 +1,63 @@
+// AWS Lambda run each request in a new process,
+// a process is not reused until a request is completed.
+//
+// Which means that each `global.context` is scoped to the lifetime of each request.
+
+// However when not in AWS Lambda, NodeJS is single-threaded, you must use the
+// per-request global context, otherwise you risk a race-condition
+// where one request overwrites another's  global context.
+//
+// Alternatively only use the local `context` in a graphql resolver.
+
 export interface GlobalContext {
-  [key: string]: any
+  [key: string]: unknown
+}
+
+import { AsyncLocalStorage } from 'async_hooks'
+
+export let asyncLocalStorage:
+  | AsyncLocalStorage<Map<string, GlobalContext>>
+  | undefined = undefined
+export const initGlobalContext = () => {
+  if (process.env.SAFE_GLOBAL_CONTEXT !== '1') {
+    asyncLocalStorage = new AsyncLocalStorage()
+    return asyncLocalStorage
+  }
+  return undefined
 }
 
-export let context: GlobalContext = {}
+let GLOBAL_CONTEXT: GlobalContext = {}
+export const context = new Proxy<GlobalContext>(GLOBAL_CONTEXT, {
```

Comment:
Nice! Not often one get to use Proxies 🙂

Reply:
They're super cool! A good candidate for `routes`!

Reply:
Since we could control the error message instead of throwing ` could not run undefined function of routes.noRealRoute()`

Reply:
Ooooor... we could redirect the user to a 404 page if the route didn't exist.

Reply:
Ohh, nice idea. I'll add it to my notes.